solr/core/src/java/org/apache/solr/cloud/api/collections/ReplicaMigrationUtils.java

```diff
@@ -0,0 +1,304 @@
+                          String.format(
+                              Locale.ROOT,
+                              "Failed to create replica for collection=%s shard=%s" + " on node=%s",
+                              sourceCollection,
+                              sourceReplica.getShard(),
+                              targetNode);
+                      log.warn(errorString);
+                      // one replica creation failed. Make the best attempt to
+                      // delete all the replicas created so far in the target
+                      // and exit
+                      synchronized (results) {
+                        results.add("failure", errorString);
+                        anyOneFailed.set(true);
+                      }
+                    } else {
+                      if (log.isDebugEnabled()) {
+                        log.debug(
+                            "Successfully created replica for collection={} shard={} on node={}",
+                            sourceCollection,
+                            sourceReplica.getShard(),
+                            targetNode);
+                      }
+                    }
+                  })
+              .get(0);
+
+      if (addedReplica != null) {
+        createdReplicas.add(addedReplica);
+        if (sourceReplica.isLeader() || waitForFinalState) {
+          String shardName = sourceReplica.getShard();
+          String replicaName = sourceReplica.getName();
+          String key = sourceCollection + "_" + replicaName;
+          CollectionStateWatcher watcher;
+          if (waitForFinalState) {
+            watcher =
+                new ActiveReplicaWatcher(
+                    sourceCollection,
+                    null,
+                    Collections.singletonList(addedReplica.getStr(ZkStateReader.CORE_NAME_PROP)),
+                    replicasToRecover);
+          } else {
+            watcher =
+                new LeaderRecoveryWatcher(
+                    sourceCollection,
+                    shardName,
+                    replicaName,
+                    addedReplica.getStr(ZkStateReader.CORE_NAME_PROP),
+                    replicasToRecover);
+          }
+          watchers.put(key, watcher);
+          log.debug("--- adding {}, {}", key, watcher);
+          ccc.getZkStateReader().registerCollectionStateWatcher(sourceCollection, watcher);
+        } else {
+          log.debug("--- not waiting for {}", addedReplica);
+        }
+      }
+    }
+
+    log.debug("Waiting for replicas to be added");
+    if (!countDownLatch.await(timeout, TimeUnit.SECONDS)) {
+      log.info("Timed out waiting for replicas to be added");
+      anyOneFailed.set(true);
+    } else {
+      log.debug("Finished waiting for replicas to be added");
+    }
+
+    // now wait for leader replicas to recover
+    log.debug("Waiting for {} leader replicas to recover", numLeaders);
+    if (!replicasToRecover.await(timeout, TimeUnit.SECONDS)) {
```

Comment:
I guess so, but I also think the logic is flawed. The replicasToRecover watch is created for the leader right after submitting the AddReplica, which can be async. So if it is async, the leader watch could count down before the replica is actually added. In general this logic is a bit iffy.

Honestly I'm not really sure why we are waiting for leaders to recover, since no nodes are being deleted here, only added. Why would the leaders go into recovery? And if so, we should wait to check for recovery until after the new replicas have been added, as I mentioned above.

Reply:
We really only need to ensure that there is another active replica to take leadership if we are deleting a leader in `cleanupReplicas()`, so we should probably just create another latch and loop for it after the `countDownLatch` is complete. (if waitForFinalState=false)

Reply:
> Honestly I'm not really sure why we are waiting for leaders to recover, since no nodes are being deleted here, only added. Why would the leaders go into recovery? And if so, we should wait to check for recovery until after the new replicas have been added, as I mentioned above.

I was thinking the same thing, I'm not sure I understand this special handling of the leaders

> We really only need to ensure that there is another active replica to take leadership if we are deleting a leader in cleanupReplicas(), so we should probably just create another latch and loop for it after the countDownLatch is complete. (if waitForFinalState=false)

Agree, I think this would be better and also make the code cleaner

Reply:
Cool, will create a different JIRA for this.